include/kframework/lemmas/int-simplification.k

```diff
@@ -115,13 +115,27 @@ module INT-SIMPLIFICATION-COMMON
   // max, min
   // ###########################################################################
 
-    rule [minint-left]: minInt(A, B) => A requires A <=Int B [simplification]
+    rule [minint-left]:  minInt(A, B) => A requires A <=Int B [simplification]
+    rule [minint-right]: minInt(A, B) => B requires B <=Int A [simplification]
 
-    rule [minint-lt]:  minInt(A, B) <Int C  => true requires A  <Int C  orBool B  <Int C [simplification]
-    rule [minint-gt]:  A  <Int minInt(B, C) => true requires A  <Int B andBool A  <Int C [simplification]
-    rule [minint-geq]: A <=Int minInt(B, C) => true requires A <=Int B andBool A <=Int C [simplification]
+    rule [minint-lt]:  minInt(A, B)  <Int C => A  <Int C  orBool B  <Int C [simplification]
+    rule [minint-leq]: minInt(A, B) <=Int C => A <=Int C  orBool B <=Int C [simplification]
+    rule [minint-gt]:  A  <Int minInt(B, C) => A  <Int B andBool A  <Int C [simplification]
+    rule [minint-geq]: A <=Int minInt(B, C) => A <=Int B andBool A <=Int C [simplification]
 
-    rule [maxint-geq-left]: A <=Int maxInt(B, C) => true requires A <=Int B orBool A <=Int C [simplification]
+    rule [minInt-factor-left]:  minInt ( A:Int +Int B:Int, A:Int +Int C:Int ) => A +Int minInt ( B, C ) [simplification]
+    rule [minInt-factor-right]: minInt ( A:Int +Int B:Int, C:Int +Int B:Int ) => minInt ( A, C ) +Int B [simplification]
+
+    rule [maxint-left]:  maxInt(A:Int, B:Int) => B requires A <=Int B [simplification]
+    rule [maxint-right]: maxInt(A:Int, B:Int) => A requires A >=Int B [simplification]
+
+    rule [maxint-lt]:  maxInt(A:Int, B:Int)  <Int C:Int => A  <Int C andBool B  <Int C [simplification]
+    rule [maxint-leq]: maxInt(A:Int, B:Int) <=Int C:Int => A <=Int C andBool B <=Int C [simplification]
+    rule [maxint-gt]:  A:Int  <Int maxInt(B:Int, C:Int) => A  <Int B  orBool A  <Int C [simplification]
+    rule [maxint-geq]: A:Int <=Int maxInt(B:Int, C:Int) => A <=Int B  orBool A <=Int C [simplification]
```

Comment:
Do we need these lemmas, or can we do with ones that go straight to `true` and `false`? In general, would like to retain as much structure of the original problem as possible. As long as one of the sides of each disjunct/conjunct is decidable as `true/false`, then these lemmas should be simplifiable, and will lead to terms that are stricly smaller. The LHS here has 2 AST nodes, and the RHS has 3, so these are expanding, and they lose the information that the disequalities came froma maxInt/minInt.

So could we instead do the splitting for each, and have only lemmas that go to `true/false`?

Reply:
Are you sure that information about where the inequalities came from would be lost? With `true/false`, we'd log the application of the simplification, but we'd do the same for these ones, so this information should ultimately be tracked, no?

When it comes to the AST structure, I do understand the point about expansion. My reasoning is that we are unconditionally eliminating a function symbol and moving to integer comparison, and so, for me, having the RHS is simpler. With `true/false`, my intuition is that we would have more conditional simplifications that would be attempted more times, which might slow the execution down.

Reply:
Sure, it's in the log, but later states don't have it, and so anything relying on `maxInt` being there won't have it, so it's no longer in the current path condition.

Same reasoning as below: either the RHS here will simplify directly down to `true/false` or one of the branches of the conjunct/disjunct _anyway_, or you'll be left with a boolean expression which gets pulled apart that came from something with higher structure (in which case, I don't really see the benefit of the boolean expression).

Using unconditional rules is good though, that does help the booster backend. But I suspect these particular conditional rules won't be too challenging for it.

Reply:
Let's go with your solution for now, sounds fine. I guess either way time will tell!